src/server.rs

```diff
@@ -704,12 +730,78 @@ impl LanguageServer for LspServer {
         };
         let pkg = parse_and_analyze(contents);
         let node = find_node(
-            walk::Node::Package(&pkg),
+            SemanticNode::Package(&pkg),
             params.text_document_position.position,
         );
 
         Ok(Some(find_references(key, node)))
     }
+    async fn completion(
+        &self,
+        params: lsp::CompletionParams,
+    ) -> RpcResult<Option<lsp::CompletionResponse>> {
+        let key =
+            params.text_document_position.text_document.uri.clone();
+
+        // We need to clone here becase `params` is used later.
+        let store = self.store.lock().unwrap().clone();
+        let contents = match store.get(&key) {
+            Some(v) => v.to_string(),
+            None => {
+                error!(
+                    "textDocument/completion called on unknown file {}",
+                    key
+                );
+                return Err(lspower::jsonrpc::Error::invalid_params(
+                    format!("file not opened: {}", key),
+                ));
+            }
+        };
+
+        let items = if let Some(ctx) = params.context.clone() {
+            match (ctx.trigger_kind, ctx.trigger_character) {
+                (
+                    lsp::CompletionTriggerKind::TriggerCharacter,
+                    Some(c),
+                ) => match c.as_str() {
+                    "." => find_dot_completions(params, contents),
+                    // XXX(sean): All `find_arg_completions` does is look for bucket names
```

Comment:
Not required, but I like sticking a date here. The worst that can happen is that we look at this comment in five years and go "HAHA this tech debt is five years old. At best, it might help us realize "Oh, this was before $FEATURE, I bet this doesn't apply anymore." I also try and keep the format `XXX: name (date) - comment`, but again, not a big deal.